package.nix

```diff
@@ -0,0 +1,384 @@
+# options to CI.
+
+, pname ? "nix"
+
+, versionSuffix ? ""
+, officialRelease ? false
+
+# Whether to build Nix. Useful to skip for tasks like (a) just
+# generating API docs or (b) testing existing pre-built versions of Nix
+, doBuild ? true
+
+# Run the unit tests as part of the build. See `installUnitTests` for an
+# alternative to this.
+, doCheck ? __forDefaults.canRunInstalled
+
+# Run the functional tests as part of the build.
+, doInstallCheck ? test-client != null || __forDefaults.canRunInstalled
+
+# Check test coverage of Nix. Probably want to use with with at least
+# one of `doCHeck` or `doInstallCheck` enabled.
+, withCoverageChecks ? false
+
+# Whether to build the regular manual
+, enableManual ? __forDefaults.canRunInstalled
+
+# Whether to compile `rl-next.md`, the release notes for the next
+# not-yet-released version of Nix in the manul, from the individual
+# change log entries in the directory.
+, buildUnreleasedNotes ? false
+
+# Whether to build the internal API docs, can be done separately from
+# everything else.
+, enableInternalAPIDocs ? false
+
+# Whether to install unit tests. This is useful when cross compiling
+# since we cannot run them natively during the build, but can do so
+# later.
+, installUnitTests ? __forDefaults.canRunInstalled
+
+# For running the functional tests against a pre-built Nix. Probably
+# want to use in conjunction with `doBuild = false;`.
+, test-daemon ? null
+, test-client ? null
+
+# Not a real argument, just the only way to approximate let-binding some
+# stuff for argument defaults.
+, __forDefaults ? {
+    canRunInstalled = doBuild && stdenv.buildPlatform.canExecute stdenv.hostPlatform;
+  }
+} @ attrs0:
+
+let
+  version = lib.fileContents ./.version + versionSuffix;
+
+  # selected attributes with defaults, will be used to define some
+  # things which should instead be gotten via `finalAttrs` in order to
+  # work with overriding.
+  attrs = {
+    inherit doBuild doCheck doInstallCheck;
+  };
+
+  filesets = {
+    baseFiles = fileset.fileFilter (f: f.name != ".gitignore") ./.;
+
+    configureFiles = fileset.unions [
+      ./.version
+      ./configure.ac
+      ./m4
+      # TODO: do we really need README.md? It doesn't seem used in the build.
+      ./README.md
+    ];
+
+    topLevelBuildFiles = fileset.unions [
+      ./local.mk
+      ./Makefile
+      ./Makefile.config.in
+      ./mk
+    ];
+
+    functionalTestFiles = fileset.unions [
+      ./tests/functional
+      (fileset.fileFilter (f: lib.strings.hasPrefix "nix-profile" f.name) ./scripts)
+    ];
+  };
+
+  mkDerivation =
+    if withCoverageChecks
+    then
+      # TODO support `finalAttrs` args function in
+      # `releaseTools.coverageAnalysis`.
+      argsFun:
+         releaseTools.coverageAnalysis (let args = argsFun args; in args)
+    else stdenv.mkDerivation;
+in
+
+mkDerivation (finalAttrs: let
+
+  inherit (finalAttrs)
+    doCheck
+    doInstallCheck
+    ;
+
+  doBuild = !finalAttrs.dontBuild;
```

Comment:
This combined with the subsequent ` dontBuild = !attrs.doBuild` feels weird. One of those seems superfluous...

Reply:
This is convenience so we are always using `do*` identifiers. They are not redundant because `attrs` vs `finalAttrs`.